src/libmongoc/src/mongoc/mongoc-client-side-encryption.c

```diff
@@ -2010,6 +2099,299 @@ mongoc_client_encryption_rewrap_many_datakey (
    RETURN (ret);
 }
 
+bool
+mongoc_client_encryption_delete_key (
+   mongoc_client_encryption_t *client_encryption,
+   const bson_value_t *keyid,
+   bson_t *reply,
+   bson_error_t *error)
+{
+   bool ret = false;
+   bson_t selector = BSON_INITIALIZER;
+
+   BSON_ASSERT_PARAM (client_encryption);
+   BSON_ASSERT_PARAM (keyid);
+   BSON_ASSERT_PARAM (reply);
+
+   BSON_ASSERT (mongoc_write_concern_get_wmajority (
+      mongoc_collection_get_write_concern (client_encryption->keyvault_coll)));
+
+   BSON_ASSERT (keyid->value_type == BSON_TYPE_BINARY);
+   BSON_ASSERT (keyid->value.v_binary.subtype == BSON_SUBTYPE_UUID);
+   BSON_ASSERT (keyid->value.v_binary.data_len > 0u);
+
+   BSON_ASSERT (BSON_APPEND_BINARY (&selector,
+                                    "_id",
+                                    keyid->value.v_binary.subtype,
+                                    keyid->value.v_binary.data,
+                                    keyid->value.v_binary.data_len));
+
+   ret = mongoc_collection_delete_one (
+      client_encryption->keyvault_coll, &selector, NULL, reply, error);
+
+fail:
+   bson_destroy (&selector);
+
+   RETURN (ret);
+}
+
+mongoc_cursor_t *
+mongoc_client_encryption_get_key (mongoc_client_encryption_t *client_encryption,
+                                  const bson_value_t *keyid,
+                                  bson_error_t *error)
+{
+   bson_t filter = BSON_INITIALIZER;
+   mongoc_cursor_t *cursor = NULL;
+
+   BSON_ASSERT_PARAM (client_encryption);
+   BSON_ASSERT_PARAM (keyid);
+
+   BSON_ASSERT (keyid->value_type == BSON_TYPE_BINARY);
+   BSON_ASSERT (keyid->value.v_binary.subtype == BSON_SUBTYPE_UUID);
+   BSON_ASSERT (keyid->value.v_binary.data_len > 0u);
+
+   BSON_ASSERT (BSON_APPEND_BINARY (&filter,
+                                    "_id",
+                                    keyid->value.v_binary.subtype,
+                                    keyid->value.v_binary.data,
+                                    keyid->value.v_binary.data_len));
+
+   BSON_ASSERT (strcmp (mongoc_read_concern_get_level (
+                           mongoc_collection_get_read_concern (
+                              client_encryption->keyvault_coll)),
+                        MONGOC_READ_CONCERN_LEVEL_MAJORITY) == 0);
+
+   /* If an error occurred, user should query cursor error. */
```

Comment:
What is the point of the `error` argument if it's never used. It looks like it might only get set when the API is disabled.

Would it make sense to copy a potential cursor error into the output parameter, or might that disturb the cursor itself?

Reply:
This was in the interest of being a lightweight wrapper around the collection operations. However, given the change in return type from `Iterable<Document>` (`mongoc_cursor_t`) to `Optional<Document>`, the cursor error (if any) should and will be stored in the `error` parameter instead.